runners/direct-java/src/main/java/org/apache/beam/runners/direct/portable/SplittableRemoteStageEvaluatorFactory.java

```diff
@@ -0,0 +1,179 @@
+import java.util.ArrayList;
+import java.util.Collection;
+import javax.annotation.Nullable;
+import org.apache.beam.model.fnexecution.v1.BeamFnApi.ProcessBundleProgressResponse;
+import org.apache.beam.model.fnexecution.v1.BeamFnApi.ProcessBundleResponse;
+import org.apache.beam.model.pipeline.v1.RunnerApi.ExecutableStagePayload;
+import org.apache.beam.runners.core.KeyedWorkItem;
+import org.apache.beam.runners.core.construction.graph.ExecutableStage;
+import org.apache.beam.runners.core.construction.graph.PipelineNode.PTransformNode;
+import org.apache.beam.runners.fnexecution.control.BundleProgressHandler;
+import org.apache.beam.runners.fnexecution.control.JobBundleFactory;
+import org.apache.beam.runners.fnexecution.control.RemoteBundle;
+import org.apache.beam.runners.fnexecution.splittabledofn.SDFFeederViaStateAndTimers;
+import org.apache.beam.runners.fnexecution.state.StateRequestHandler;
+import org.apache.beam.runners.fnexecution.wire.WireCoders;
+import org.apache.beam.sdk.coders.Coder;
+import org.apache.beam.sdk.coders.KvCoder;
+import org.apache.beam.sdk.transforms.windowing.BoundedWindow;
+import org.apache.beam.sdk.util.WindowedValue;
+import org.apache.beam.sdk.util.WindowedValue.FullWindowedValueCoder;
+import org.apache.beam.sdk.values.KV;
+
+/**
+ * The {@link TransformEvaluatorFactory} for {@link #URN}, which reads from a {@link
+ * DirectGroupByKey#DIRECT_GBKO_URN} and feeds the data, using state and timers, to a {@link
+ * ExecutableStage} whose first instruction is an SDF.
+ */
+class SplittableRemoteStageEvaluatorFactory implements TransformEvaluatorFactory {
+  public static final String URN = "urn:beam:directrunner:transforms:splittable_remote_stage:v1";
+
+  // A fictional transform that transforms from KWI<unique key, KV<element, restriction>>
+  // to simply KV<element, restriction> taken by the SDF inside the ExecutableStage.
+  public static final String FEED_SDF_URN = "urn:beam:directrunner:transforms:feed_sdf:v1";
+
+  private final BundleFactory bundleFactory;
+  private final JobBundleFactory jobBundleFactory;
+  private final StepStateAndTimers.Provider stp;
+
+  SplittableRemoteStageEvaluatorFactory(
+      BundleFactory bundleFactory,
+      JobBundleFactory jobBundleFactory,
+      StepStateAndTimers.Provider stepStateAndTimers) {
+    this.bundleFactory = bundleFactory;
+    this.jobBundleFactory = jobBundleFactory;
+    this.stp = stepStateAndTimers;
+  }
+
+  @Nullable
+  @Override
+  public <InputT> TransformEvaluator<InputT> forApplication(
+      PTransformNode application, CommittedBundle<?> inputBundle) throws Exception {
+    return new SplittableRemoteStageEvaluator(
+        bundleFactory,
+        jobBundleFactory,
+        stp.forStepAndKey(application, inputBundle.getKey()),
+        application);
+  }
+
+  @Override
+  public void cleanup() throws Exception {
+    jobBundleFactory.close();
+  }
+
+  private static class SplittableRemoteStageEvaluator<InputT, RestrictionT>
```

Comment:
In a future PR I could see this completely replacing RemoteStageEvaluator for non SDF ExecutableStages and do a no-op if there are no splits if we can figure out a good way to handle the small bit of code which is needed to support knowing that we are resuming an SDF. It doesn't make sense yet because we don't yet know how dynamic work rebalancing will impact this code.